phd-tests/framework/src/disk/fat.rs

```diff
@@ -0,0 +1,209 @@
+/// clusters it has, but the number of clusters in the volume depends on the
+/// size of the filesystem metadata, which depends on the volume's format.
+/// This module avoids the problem by requiring that the total volume size is
+/// less than or equal to 4,084 sectors. The number of clusters will never be
+/// greater than this, so the filesystem will always be FAT12, which makes it
+/// possible to compute its overhead size.
+const fn overhead_sectors() -> Sectors {
+    let dir_entry_bytes = ROOT_DIRECTORY_ENTRIES * BYTES_PER_DIRECTORY_ENTRY;
+    let dir_entry_sectors = Sectors::needed_for_bytes(dir_entry_bytes);
+
+    // To compute the size of the FAT, conservatively assume that all the
+    // sectors on the disk are in addressable clusters, figure out how many
+    // bytes that would take, and convert to sectors. (In practice, some of the
+    // sectors are used for overhead and won't have entries in the FAT, but this
+    // gives an upper bound.)
+    let max_sectors = Sectors::needed_for_bytes(MAX_DISK_SIZE_BYTES);
+
+    // FAT12 tables use 12 bits per cluster entry.
+    let cluster_bits = max_sectors.0 * 12;
+    let cluster_bytes = cluster_bits.div_ceil(8);
+    let sectors_per_table = Sectors::needed_for_bytes(cluster_bytes);
+
+    // The total overhead size is one sector (for the BPB) plus the sectors
+    // needed for regions 2 and 3 (the tables themselves and the root directory
+    // entries). Note that there are multiple tables per volume!
+    Sectors(1 + (TABLES_PER_VOLUME * sectors_per_table.0) + dir_entry_sectors.0)
+}
+
+const fn usable_sectors() -> Sectors {
+    let total = Sectors::needed_for_bytes(MAX_DISK_SIZE_BYTES);
+    Sectors(total.0 - overhead_sectors().0)
+}
+
+#[derive(Clone, Debug)]
+struct File {
+    path: String,
+    contents: Vec<u8>,
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error(
+        "insufficient space for new file: {0} sectors required, {1} available"
+    )]
+    NoSpace(usize, usize),
+}
+
+/// Builds a collection of files that can be extruded as an array of bytes
+/// containing a FAT filesystem with the collected files.
+#[derive(Clone, Default, Debug)]
+pub struct FatFilesystem {
+    files: Vec<File>,
+    sectors_used: Sectors,
+}
+
+impl FatFilesystem {
+    /// Creates a new file collection.
+    pub fn new() -> Self {
+        Self { files: vec![], sectors_used: Sectors(0) }
+    }
+
+    /// Adds a file with the supplied `contents` that will appear in the
+    /// resulting file system in the supplied `path`.
+    ///
+    /// N.B. `path` is interpreted relative to the root of the file system and
+    ///      should not have a leading '/'.
+    pub fn add_file_from_str(
+        &mut self,
+        path: &str,
+        contents: &str,
+    ) -> Result<(), Error> {
+        let bytes = contents.as_bytes();
+        let sectors_needed = Sectors::needed_for_bytes(bytes.len());
+        let sectors_available = usable_sectors().0 - self.sectors_used.0;
```

Comment:
is it possible for `self.sectors_used` to have exceeded `usable_sectors()` here? I couldn't find the check for that...

Reply:
It's not supposed to be (any addition that would cause that post-condition should return `NoSpace` instead), but I agree that this wasn't very obvious, so I flipped the polarity of the tracker in bccf54b, which I think improves things.